fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinOrder.java

```diff
@@ -294,39 +293,47 @@ protected void computeCost(ExpressionInfo exprInfo) {
     protected ExpressionInfo buildJoinExpr(GroupInfo leftGroup, GroupInfo rightGroup) {
         ExpressionInfo leftExprInfo = leftGroup.bestExprInfo;
         ExpressionInfo rightExprInfo = rightGroup.bestExprInfo;
-        Pair<ScalarOperator, ScalarOperator> predicates = buildInnerJoinPredicate(
-                leftGroup.atoms, rightGroup.atoms);
+        Pair<ScalarOperator, ScalarOperator> predicates = buildInnerJoinPredicate(leftGroup.atoms, rightGroup.atoms);
+
+        ScalarOperator eqPredicate = predicates.first;
+        ScalarOperator otherPredicate = predicates.second;
+
         LogicalJoinOperator newJoin;
-        if (predicates.first != null) {
-            newJoin =
-                    new LogicalJoinOperator(JoinOperator.INNER_JOIN, predicates.first);
+        if (eqPredicate != null) {
+            newJoin = new LogicalJoinOperator(JoinOperator.INNER_JOIN, eqPredicate);
         } else {
-            newJoin =
-                    new LogicalJoinOperator(JoinOperator.CROSS_JOIN, null);
+            newJoin = new LogicalJoinOperator(JoinOperator.CROSS_JOIN, null);
         }
-        newJoin.setPredicate(predicates.second);
+        newJoin.setPredicate(otherPredicate);
 
         Map<ColumnRefOperator, ScalarOperator> leftExpression = new HashMap<>();
         Map<ColumnRefOperator, ScalarOperator> rightExpression = new HashMap<>();
-        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : expressionMap.entrySet()) {
-            // If entry.getValue is Constant, then this ColumnRef does not belong to any child.
-            // Then you can add this constant mapping on any child
-            if (predicates.first != null && predicates.first.getUsedColumns().contains(entry.getKey())) {
-                if (leftExprInfo.expr.getOutputColumns().containsAll(entry.getValue().getUsedColumns())
-                        || entry.getValue() instanceof ConstantOperator) {
+        if (eqPredicate != null || otherPredicate != null) {
+            ColumnRefSet useColumns = new ColumnRefSet();
+
+            if (eqPredicate != null) {
+                useColumns.union(eqPredicate.getUsedColumns());
+            }
+
+            if (otherPredicate != null) {
+                useColumns.union(otherPredicate.getUsedColumns());
+            }
+
+            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : expressionMap.entrySet()) {
+                if (!useColumns.contains(entry.getKey())) {
+                    continue;
+                }
+
+                if (entry.getValue().isConstantRef()) {
+                    // constant always on left
                     leftExpression.put(entry.getKey(), entry.getValue());
```

Comment:
Is there improve the performance significantly?

Reply:
can be twice as fast on join reorder。。。